llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -4147,6 +4147,22 @@ AArch64TTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *ValTy,
   switch (ISD) {
   default:
     break;
+  case ISD::FADD: {
+    if (MTy.isVector()) {
+      // FIXME: Consider cases where the number of vector elements is not power
+      // of 2.
+      const unsigned NElts = MTy.getVectorNumElements();
+      if (ValTy->getElementCount().getFixedValue() >= 2 && NElts >= 2 &&
+          isPowerOf2_32(NElts)) {
```

Comment:
Can you check the fp type is one that we would expect. I think the MTy.isVector() is protecting against fp128 but those should have a high cost too. The rule for fp16 should generally be that if +fullfp16 is present then it is cheap, otherwise it needs to extend it to fp32 and reduce that.

Reply:
> Can you check the fp type is one that we would expect. I think the MTy.isVector() is protecting against fp128 but those should have a high cost too. The rule for fp16 should generally be that if +fullfp16 is present then it is cheap, otherwise it needs to extend it to fp32 and reduce that.

I assume you are stating the case where input vector is v8f16 (i.e. fp128).

Isn't taking into account legalization cost here sufficient for the case you have cited?
```
return (LT.first - 1) + /*No of faddp instructions*/ Log2_32(NElts)
```

Reply:
The cost without fp16:
```
; CHECK-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %fadd_v8f16 = call fast half @llvm.vector.reduce.fadd.v8f16(half 0xH0000, <8 x half> undef)
```
seems to be the same as with fp16, I would expect it to be higher.
```
; FP16-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %fadd_v8f16 = call fast half @llvm.vector.reduce.fadd.v8f16(half 0xH0000, <8 x half> undef)
```
It might be that the type legalization cost doesn't account for it as the type is still legal to some extent.

Reply:
> The cost without fp16:
> 
> ```
> ; CHECK-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %fadd_v8f16 = call fast half @llvm.vector.reduce.fadd.v8f16(half 0xH0000, <8 x half> undef)
> ```
> 
> seems to be the same as with fp16, I would expect it to be higher.
> 
> ```
> ; FP16-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %fadd_v8f16 = call fast half @llvm.vector.reduce.fadd.v8f16(half 0xH0000, <8 x half> undef)
> ```
> 
> It might be that the type legalization cost doesn't account for it as the type is still legal to some extent.

ok, I will add a check:
```
if( type == half  && fp16enabled() )
    return <patch cost>
else
    fallback to current cost
```
    
does the above check look good?

Reply:
Yeah that sounds OK. We have some other code for fneg costs that does this and checks fp32 and fp64 too, but some others are a little different:
```
if (Ty->isFloatTy() || Ty->isDoubleTy() || (Ty->isHalfTy() && ST->hasFullFP16()))) 
  do patch cost
else
  fallback to current cost 
```

Reply:
@davemgreen sorry, got bit confused with the types above(and hence your initial suggestion).

For half type:
I agree `-fullfp16` cost should be higher than `+fullfp16`. But don't you think thats a different issue altogether and needs to be taken differently from this patch?

for fp128 type:
why we should not allow fp128? are there any issues with this type? 
My understanding is fp128 is multiple of fp32 and it also occupies whole register and hence, no legalization cost would be involved. Correct me if wrong.

Reply:
@davemgreen thanks for input.